enhancements/insights/pulling-data-from-ocm.md

```diff
@@ -0,0 +1,133 @@
+---
+title: pulling-and-exposing-data-from-ocm
+authors:
+  - "@tremes"
+reviewers:
+  - "@sbose78"
+  - "@inecas"
+  - "@petli-openshift"
+  - "@smarterclayton"
+approvers:
+  - "@sbose78"
+  - "@smarterclayton"
+creation-date: 2021-03-04
+last-updated: 2021-03-09
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Insights Operator pulling and exposing data from the OCM API
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement will enable the Insights Operator to pull the data (SCA certs)
+from the OCM (OpenShift Cluster Manager) API. The data will be exposed by the Insights Operator
+in the OpenShift API to allow users to use them when consuming and building container images
+on the platform.
+
+## Motivation
+
+Users could consume RHEL content and container images using the RHEL subscription in the OpenShift 3.x.
+In the OpenShift 4, this is no longer possible because the Red Hat Enterprise Linux Core OS (RHCOS) does not
+provide any attached subscription. This enhancement is to provide users the Simple Content Access (SCA) certs
+from Red Hat Subscription Manager (RHSM).
+
+### Goals
+
+- Extend the Insights Operator config with an OCM API URL to be able to query the data
+- Periodically pull the data from the OCM API and expose it in the OpenShift API
+
+### Non-Goals
+
+- Insights Operator providing any transformation or post-processing of the data pulled
+  from the OCM API
+
+## Proposal
+
+### User Stories
+
+#### Consume SCA certs exposed in the API
+
+As an OpenShift user
+I want to consume SCA certs to be able to consume RHEL content and to build
+corresponding container images.
+
+### Risks and Mitigations
+
+#### OCM API is down
+
+Risk: OCM API is down or doesn't provide up to date data.
+
+Risk: Insights Operator is unable to expose/update the data in the OpenShift API
+
+Mitigation: Introduce a new state in the Insights Operator (e.g "SCADataDegraded") and
+create a new alert based on this new state.
+
+## Design Details
+
+### Authorization
+
+The Insights Operator is able to pull the data from the OCM API using the existing `cloud.openshift.com` token
+available in the `pull-secret` (in the `openshift-config-managed` namespace).
+
+The Insights Operator must provide a cluster ID as an identifier of the cluster.
+
+### Data in API
+
+The SCA certificate is available via the `etc-pki-entitlement` secret in the `openshift-config-managed` namespace.
```

Comment:
how is this certificate going to be injected into a pod?

Reply:
I am not sure it's the goal of this proposal to answer this question. Do we need to answer it here? @sbose78 is working on "tooling" using this cert.

Reply:
The Insights Operator will also need to do the following:

1. Create a `Share` object for the projected resource CSI driver
2. Create a "viewer" ClusterRole that allows read access to the Share (get, list, watch)

Ideally we aggregate this cluster role - I would recommend aggregation to the k8s admin and edit roles so end users and our default service accounts can access it [1].

Pods will then need to mount the share via a CSI inline volume - see openshift/csi-driver-projected-resource for an example [2].

[1] https://kubernetes.io/docs/reference/access-authn-authz/rbac/#user-facing-roles
[2] https://github.com/openshift/csi-driver-projected-resource/blob/master/docs/simple-example.md

Reply:
@adambkaplan Can you please share more details or requirements for the `Share` object and the ClusterRole. I guess the IO has to define a `Share` object where the `backingResource` is the new secret right? And the ClusterRole is then able to read this `Share` object right?

Reply:
Historical note: I had agreed to have de-scope creation of the `Share` object(and related RBAC) since we wanted to make progress with getting the SCA cert into the cluster and provide a path to admins to make it available for users.

Reply:
> I am not sure it's the goal of this proposal to answer this question.

i think it's worth mentioning the expectations (because those expectations tie into how you intend to secure the content, where it should be stored, as well as demonstrating that there is value in doing this and that it's being done in a way that those downstream consumers will find useful), but i agree that the answer should be general and not get into specific implementation details.

Reply:
This is described in the [Use of the SCA certs](https://github.com/openshift/enhancements/pull/683/files#diff-93dffd2aa59c973a916306675f62907275cff29ff9a16ba77aaf98cfa830249fR97) section.